src/GraphQL.Tests/Complexity/ComplexityValidationTests.cs

```diff
@@ -104,4 +106,70 @@ public void fail_when_too_complex_and_nested()
         res.Result.Errors[0].Message.ShouldBe("Query is too complex to execute. Complexity is 480, maximum allowed on this endpoint is 25. The field with the highest complexity is 'friends' with value 125.");
         res.Result.Errors[0].InnerException.ShouldBeNull();
     }
+
+    // https://github.com/graphql-dotnet/graphql-dotnet/issues/3527
+    [Fact]
+    public void recursive_fragment_should_not_end_in_eternal_loop1()
+    {
+        const string query = """
+            {
+                type_All(limit: 20) {
+                    items {
+                        links(limit: 20) {
+                            items {
+                                ... RecursiveFragment
+                            }
+                        }
+                    }
+                }
+            }
+            fragment RecursiveFragment on Type   {
+               links(limit: 20) {
+                    items {
+                        ... RecursiveFragment
+                    }
+                }
+            }
+            """;
+
+        var complexityConfiguration = new ComplexityConfiguration();
+        var res = Execute(complexityConfiguration, query);
+
+        res.Result.Errors.ShouldNotBe(null);
+        res.Result.Errors.Count.ShouldBe(3);
+        res.Result.Errors.All(e => e is not ComplexityError);
+    }
+
+    // https://github.com/graphql-dotnet/graphql-dotnet/issues/3527
+    [Fact]
+    public void recursive_fragment_should_not_end_in_eternal_loop2()
+    {
+        const string query = """
+            {
+                type_All(limit: 20) {
+                    items {
+                        links(limit: 20) {
+                            items {
+                                ... RecursiveFragment
+                            }
+                        }
+                    }
+                }
+            }
+            fragment RecursiveFragment on Type   {
+               links(limit: 20) {
+                    items {
+                        ... RecursiveFragment
+                    }
+                }
+            }
+            """;
+
+        var complexityConfiguration = new ComplexityConfiguration();
+        var res = Execute(complexityConfiguration, query, onlyComplexityRule: true);
```

Comment:
This should test with all rules and verify that only one error is produced, and no deadlock occurs.

Reply:
`recursive_fragment_should_not_end_in_eternal_loop1` is the one with all rules

Reply:
I guess reviewing from my cell phone is proving difficult….!!!

Reply:
I review only the simplest PRs from cell phone. It is simply unrealistic to work with more serious edits.